File: nhglib-core/src/io/github/voidzombie/nhglib/graphics/lights/tiled/NhgLightsAttribute.java
package io.github.voidzombie.nhglib.graphics.lights.tiled;

import com.badlogic.gdx.graphics.g3d.Attribute;
import com.badlogic.gdx.utils.Array;

/**
 * Created by Fausto Napoli on 18/03/2017.
 */
public class NhgLightsAttribute extends Attribute {
    public final static String Alias = "nhgLights";
    public final static long Type = register(Alias);

    public final static boolean is(final long mask) {
        return (mask & Type) == mask;
    }

    public final Array<NhgLight> lights;

    public NhgLightsAttribute() {
        super(Type);
        lights = new Array<>(1);
    }

    public NhgLightsAttribute(final NhgLightsAttribute copyFrom) {
        this();
        lights.addAll(copyFrom.lights);
    }

    @Override
    public NhgLightsAttribute copy() {
        return new NhgLightsAttribute(this);
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        for (NhgLight light : lights)
            result = 1237 * result + (light == null ? 0 : light.hashCode());
        return result;
    }

    @Override
    public int compareTo(Attribute o) {
        if (type != o.type) return type < o.type ? -1 : 1;
        return 0; // FIXME implement comparing
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/graphics/lights/NhgLightsAttribute.kt
package io.github.movementspeed.nhglib.graphics.lights

import com.badlogic.gdx.graphics.g3d.Attribute
import com.badlogic.gdx.utils.Array

/**
 * Created by Fausto Napoli on 18/03/2017.
 */
class NhgLightsAttribute() : Attribute(Type) {

    val lights: Array<NhgLight>

    init {
        lights = Array(1)
    }

    constructor(copyFrom: NhgLightsAttribute) : this() {
        lights.addAll(copyFrom.lights)
    }

    override fun copy(): NhgLightsAttribute {
        return NhgLightsAttribute(this)
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        for (light in lights)
            result = 1237 * result + (light?.hashCode() ?: 0)
        return result
    }

    override fun compareTo(o: Attribute): Int {
        return if (type != o.type) if (type < o.type) -1 else 1 else 0
    }

    companion object {
        val Alias = "nhgLights"
        val Type = Attribute.register(Alias)

        fun `is`(mask: Long): Boolean {
            return mask and Type == mask
        }
    }
}
